src/JsonApiDotNetCore.MongoDb/MongoDbRepository.cs

```diff
@@ -0,0 +1,230 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using JsonApiDotNetCore.Configuration;
+using JsonApiDotNetCore.Queries.Internal.QueryableBuilding;
+using JsonApiDotNetCore.Repositories;
+using JsonApiDotNetCore.Resources;
+using MongoDB.Driver;
+using MongoDB.Driver.Linq;
+using JsonApiDotNetCore.Queries;
+using JsonApiDotNetCore.Queries.Expressions;
+using Microsoft.EntityFrameworkCore.Metadata;
+using Microsoft.EntityFrameworkCore.Infrastructure;
+
+namespace JsonApiDotNetCore.MongoDb
+{
+    /// <summary>
+    /// Implements the foundational Repository layer in the JsonApiDotNetCore architecture that uses MongoDB.
+    /// </summary>
+    public class MongoDbRepository<TResource, TId>
+        : IResourceRepository<TResource, TId>
+        where TResource : class, IIdentifiable<TId>
+    {
+        private readonly IMongoDatabase _mongoDatabase;
+        private readonly ITargetedFields _targetedFields;
+        private readonly IResourceContextProvider _resourceContextProvider;
+        private readonly IResourceFactory _resourceFactory;
+        
+        public MongoDbRepository(
+            IMongoDatabase mongoDatabase,
+            ITargetedFields targetedFields,
+            IResourceContextProvider resourceContextProvider,
+            IResourceFactory resourceFactory)
+        {
+            _mongoDatabase = mongoDatabase ?? throw new ArgumentNullException(nameof(mongoDatabase));
+            _targetedFields = targetedFields ?? throw new ArgumentNullException(nameof(targetedFields));
+            _resourceContextProvider = resourceContextProvider ?? throw new ArgumentNullException(nameof(resourceContextProvider));
+            _resourceFactory = resourceFactory ?? throw new ArgumentNullException(nameof(resourceFactory));
+        }
+
+        protected virtual IMongoCollection<TResource> Collection => _mongoDatabase.GetCollection<TResource>(typeof(TResource).Name);
+
+        /// <inheritdoc />
+        public virtual async Task<IReadOnlyCollection<TResource>> GetAsync(QueryLayer layer,
+            CancellationToken cancellationToken)
+        {
+            if (layer == null) throw new ArgumentNullException(nameof(layer));
+            var list = await ApplyQueryLayer(layer).ToListAsync(cancellationToken);
+            return list.AsReadOnly();
```

Comment:
Please add a blank line after input validation. Also a better name for variable `list` (the type) would be `resources` (what it represents).